### PR TITLE
Avoid repeatedly instantiating potentials

### PR DESCRIPTION
**Pull request by badisa**
* Re-using the preallocated potentials appears to get us back some performance

# Benchmarks
The reference refers to what is currently on master, the comparison is this branch. Generated using the test here: https://github.com/proteneer/timemachine/blob/1b8f07f102cdeab76f5b7e2126dcd8ba7bfcc49e/tests/test_benchmark_free_energy.py#L76 with the additional change to run with 2, 4, 8, 16, 32 and 48 windows as well as making sure every window always ran 100 frames. 

Run on an A10 GPU with Cuda Arch 8.6. Only sampled once, but think the differences are significant even without replicates.

Also ran a small graph (bisection + hrex + water sampling) that was about 5% faster with this change. 

## Vacuum
Doesn't make a difference on vacuum simulations, presumably the cost of constructing potentials is cheap (no neighborlist notably, tiny systems, etc)

image

## Solvent
Seems to make the biggest difference for bisection

image

### Without water sampling
image

### With water sampling
image

### CODE REVIEW

**Review by maxentile — APPROVED** [review]
Changes look good to me!

Seeing similar timings locally with `pytest -k "test_benchmark_hif2a_single_topology"`

> Generated using the test here:
> ...
> with the additional change to run with 2, 4, 8, 16, 32 and 48 windows as well as making sure every window always ran 100 frames.

Tangent: The timings for each of these 3 modes at this grid of n_windows is very informative.

Would you mind adding these changes (+ either the plot or some other convenient export of the resulting timings) to test_benchmark_free_energy.py ? Will be good to track timings across this grid of conditions in the future.

A few observations about current timings. These are not necessarily surprising (most make sense in context), and are not immediately actionable, but they do deviate from some ideal:
* throughput for vacuum simulations (~6000-8000 ns/day) is only ~10x higher than for complex (~600-700 ns/day), even though the particle count is ~1000x smaller. (and only ~6x higher than for solvent (~900-1000 ns/day), even though the particle count is ~250x smaller)
* throughput decreases or stays constant with the number of windows. (in principle, should be able to achieve the same or higher throughput, since we increase the number of independent tasks to perform.)
* hrex overhead appears approximately linear in # windows. (Eyeballing: maybe a penalty of ~3-5 ns/day per window, relative to independent simulations?) Presumably, computing `u_kl` is the dominant cost, but I haven't verified.

Oddity: (probably inconsequential, just surprising)
* Vacuum sequential timings have an anomaly for small # windows ([2,4,8] windows, [~8000, ~5800, ~6200] ns/day)

**Comment by badisa** [comment]
> Would you mind adding these changes (+ either the plot or some other convenient export of the resulting timings) to test_benchmark_free_energy.py ? Will be good to track timings across this grid of conditions in the future.
>

Added in https://github.com/proteneer/timemachine/pull/1287/commits/19b5c7d655e32c6ad2b3492717f0021dc9591298

**Comment by badisa** [comment]
Out of curiosity I ran vacuum and solvent over night going out to 2000 frames to see if the performance differences changed. Only noticeable difference was that with solvent HREX only ~20% of ns/day was lost versus 65% lost in vacuum.  Which makes sense as solvent is more likely to be dominated by the cost of MD rather than by the cost of whatever work these methods do on their own.  

vacuum_2000_benchmarks
solvent_2000_benchmarks

File: timemachine/fe/energy_decomposition.py [hunk 1]
```diff
@@ -37,17 +42,26 @@ def get_batch_u_fns(bps: Iterable[BoundPotential], temperature: float = DEFAULT_
     """
     kBT = temperature * BOLTZ
 
+    assert len(pots) == len(params)
     batch_u_fns: List[Batch_u_fn] = []
-    for bp in bps:
+    for p, bp in zip(params, pots):
```
Comment (mcwitt) [inline timemachine/fe/energy_decomposition.py:47]: nit: should we use a name other than `bp` for unbound potentials?

Comment (badisa) [reply timemachine/fe/energy_decomposition.py:47]: Good catch, will make it more explicit. Do we have a convention like we do for bound potentials?

File: timemachine/fe/free_energy.py [hunk 2]
```diff
@@ -1141,19 +1137,25 @@ def format_rates(rs):
...
+    assert isinstance(potential, custom_ops.SummedPotential)
+    unbound_impls = potential.get_potentials()
+
     def make_energy_decomposed_state(
-        results: Tuple[StoredArrays, List[NDArray], InitialState]
+        results: Tuple[StateIdx, StoredArrays, List[NDArray]]
     ) -> EnergyDecomposedState[StoredArrays]:
-        frames, boxes, initial_state = results
+        state_idx, frames, boxes = results
+        initial_state = initial_states[state_idx]
```
Comment (mcwitt) [inline timemachine/fe/free_energy.py:1149]: Curiosity: what was the motivation for the change to pass in a `StateIdx` and use it to look up the `initial_state`, versus directly passing in the `InitialState` as before?

File: timemachine/fe/free_energy.py [hunk 3]
```diff
@@ -728,6 +726,7 @@ def run_sims_sequential(
     # u_kln matrix (2, 2, n_frames) for each pair of adjacent lambda windows and energy term
     u_kln_by_component_by_lambda = []
 
+    unbound_impls = [p.potential.to_gpu(np.float32).unbound_impl for p in initial_states[0].potentials]
```
Comment (mcwitt) [inline timemachine/fe/free_energy.py:729]: Would it be worth adding a warning that this assumes all initial states have the same potentials (as [here](https://github.com/proteneer/timemachine/blob/7f1bffc7f24fd1d167b43608f1f1a0f6ae1a07fa/timemachine/fe/free_energy.py#L986))?

File: timemachine/fe/energy_decomposition.py (hunk 1 above)
Comment (mcwitt) [reply timemachine/fe/energy_decomposition.py:47]: I'm not aware of a convention. I think I've mostly used `pot` or `potential` (or just `p` if the scope is local and no potential for confusion with parameters)

**Review by mcwitt — APPROVED** [review]
Looks great! Nice perf improvement!

File: timemachine/fe/energy_decomposition.py (hunk 1 above)
Comment (badisa) [reply timemachine/fe/energy_decomposition.py]: Fixed up in all the places I could find: [1eff706](https://github.com/proteneer/timemachine/pull/1287/commits/1eff706be5cba8fe393e0eb2fb333d42c405a3da)

File: timemachine/fe/free_energy.py (hunk 2 above)
Comment (badisa) [reply timemachine/fe/free_energy.py]: Reverted to reduce diff: [da0f7f8](https://github.com/proteneer/timemachine/pull/1287/commits/da0f7f8f29d9c2576dc31bee7b49351d05e7cb28)

**Pull request merged**